sw/cheri/common/sdcard-utils.hh

```diff
@@ -0,0 +1,472 @@
+        (void)get_response_R1();
+      }
+
+      if (!collected_data(&buf[blk * kBlockLen], kBlockLen)) {
+        ok = false;
+        break;
+      }
+    }
+
+    if (multi) {
+      send_command(CMD_STOP_TRANSMISSION, 0u);
+      (void)get_response_R1b();
+    }
+
+    select_card(false);
+    return ok;
+  }
+
+  // Write a number of contiguous blocks from the SD card.
+  // TODO: Support non-blocking operation.
+  bool write_blocks(uint32_t block, uint8_t *buf, size_t num_blocks = 1u, bool blocking = true) {
+    const bool multi = num_blocks > 1u;
+    uint8_t crc16[2];
+    crc16[1] = crc16[0] = 0xffu;  // CRC16 not required by default for SPI mode.
+
+    select_card(true);
+
+    bool ok(true);
+    for (size_t blk = 0u; blk < num_blocks; blk++) {
+      if (crcOn) {
+        // CRC16 bytes follow the data block.
+        uint16_t crc = calc_crc16(&buf[blk * kBlockLen], kBlockLen);
+        crc16[0]     = (uint8_t)(crc >> 8);
+        crc16[1]     = (uint8_t)crc;
+      }
+      if (log) {
+        log->println("Writing block {}", block + blk);
+      }
+
+      // Note: the Start Block Token differs between Multiple Block Write commands and
+      // the other data transfer commands, including the Single Block Write command.
+      uint8_t start_token;
+      if (multi) {
+        // Is this the first block of the read request?
+        if (!blk) {
+          send_command(SdCard::CMD_WRITE_MULTIPLE_BLOCK, block);
+          (void)get_response_R1();
+        }
+        start_token = StartBlockTokenMW;
+      } else {
+        send_command(SdCard::CMD_WRITE_SINGLE_BLOCK, block + blk);
+        (void)get_response_R1();
+        start_token = StartBlockToken;
+      }
+
+      spi->blocking_write(&start_token, 1u);
+      spi->blocking_write(&buf[blk * kBlockLen], kBlockLen);
+      spi->blocking_write(crc16, sizeof(crc16));
+      // Collect data_response and wait until the card is no longer busy.
+      if (5 != (0x1f & get_data_response_busy())) {
+        // Data not accepted because of an error.
+        ok = false;
+        break;
+      }
+    }
+
+    if (multi) {
+      const uint8_t stop_tran_token = (uint8_t)StopTranToken;
+      spi->blocking_write(&stop_tran_token, 1u);
+      // The card will hold the CIPO line low whilst busy, yielding repeated 0x00 bytes,
+      // but it seems to drop and raise the line at an arbitrary time with respect to
+      // the '8-clock counting' logic.
+      while (0x00 != get_response_byte());  // Detect falling edge.
+      // Card will signal busy with zeros.
+      wait_not_busy();
+    }
+
+    select_card(false);
+    return ok;
+  }
+
+  // Send a command to the SD card with the supplied 32-bit argument.
+  void send_command(uint8_t cmdCode, uint32_t arg) {
+    uint8_t cmd[6];
+    if (log) {
+      log->println("Sending command {:#04x}", cmdCode);
+    }
+
+    // Apparently we need to clock 8 times before sending the command.
```

Comment:
This is actually the opposite: you need 8 clock ticks *after commands*. Per the spec:

> After the last SD Memory Card bus transaction, the host is required, to provide 8 (eight) clock
> cycles for the card to complete the operation before shutting down the clock. Following is a list of the
> various bus transactions:
> •A command with no response. 8 clocks after the host command end bit.
> •A command with response. 8 clocks after the card response end bit.
> •A read data transaction. 8 clocks after the end bit of the last data block.
> •A write data transaction. 8 clocks after the CRC status token.